### PR TITLE
Capture and log page errors

### PR DESCRIPTION
**Pull request by shahzad31**
Fixes https://github.com/elastic/synthetics/issues/226

Add an event handler in browser console for "pageerror".

Capturing `error.message` and `error.stack.toString()`

### CODE REVIEW

File: src/plugins/browser-console.ts [hunk 1]
```diff
@@ -54,12 +54,31 @@ export class BrowserConsole {
...
+    if (!this._currentStep) {
+      return;
+    }
+    const { name, index } = this._currentStep;
+    this.messages.push({
+      timestamp: getTimestamp(),
+      text: error.message,
+      type: 'error',
+      step: { name, index },
+      stacktrace: error.stack.toString(),
```
Comment (vigneshshanmugam) [inline src/plugins/browser-console.ts:67]: pass it as error, we have a global formatError function that handles them properly in the console event - https://github.com/elastic/synthetics/blob/e67a91dacc4c981aeb6b0f9d715f07b059da8f80/src/reporters/json.ts#L486-L495

File: src/plugins/browser-console.ts [hunk 2]
```diff
@@ -54,12 +54,31 @@ export class BrowserConsole {
...
+    });
+    if (this.messages.length > defaultMessageLimit) {
```
Comment (vigneshshanmugam) [inline src/plugins/browser-console.ts:69]: Move this logic to stop to avoid doing at multiple places.

File: src/common_types.ts [hunk 3]
```diff
@@ -119,6 +119,7 @@ export type NetworkInfo = {
 export type BrowserMessage = {
   text: string;
   type: string;
+  stacktrace?: string;
```
Comment (vigneshshanmugam) [inline src/common_types.ts:122]: would become error, see the formatError comment.

**Review by vigneshshanmugam — COMMENTED** [review]
Thanks for working on this, tests can be written in `browser-console.test.ts`, where you can throw some promise errors inside the evaluate function. 
```
    await page.evaluate(() => Promise.reject(new Error("Boom")));
    // make sure to add some stack here and do the needful. 
```

Let me know if you need any help.

**Comment by shahzad31** [comment]
@vigneshshanmugam i have added a unit test

File: src/reporters/json.ts [hunk 4]
```diff
@@ -483,13 +483,17 @@ export default class JSONReporter extends BaseReporter {
...
         if (browserconsole) {
-          browserconsole.forEach(({ timestamp, text, type, step }) => {
+          browserconsole.forEach(({ timestamp, text, type, step, error }) => {
             this.writeJSON({
               type: 'journey/browserconsole',
               journey,
               timestamp,
               step,
-              payload: { text, type } as Payload,
+              payload: {
```
Comment (vigneshshanmugam) [inline src/reporters/json.ts:492]: you can pass error as a standalone field here. will be formatted correctly in write handler.

File: __tests__/plugins/browser-console.test.ts [hunk 5]
```diff
@@ -55,4 +55,49 @@ describe('BrowserConsole', () => {
...
+  it('should capture browser page errors', async () => {
+    const driver = await Gatherer.setupDriver({ wsEndpoint });
+    const browserConsole = new BrowserConsole(driver);
+    const { page } = driver;
+    browserConsole.start();
+    await page.goto(server.TEST_PAGE);
+    browserConsole._currentStep = { name: 'step-name', index: 0 };
+    const bodyHandle = await page.$('body');
+    try {
+      await page.evaluate(
```
Comment (vigneshshanmugam) [inline __tests__/plugins/browser-console.test.ts:68]: You can do await page.setContent to make this simpler.

File: __tests__/plugins/browser-console.test.ts [hunk 6]
```diff
@@ -55,4 +55,49 @@ describe('BrowserConsole', () => {
...
+    const messages = browserConsole.stop();
+    await Gatherer.stop();
+    const notFoundMessage = messages.find(
+      m => m.text.indexOf('Failed to load resource:') >= 0
+    );
+    expect(notFoundMessage.text).toEqual(
+      `Failed to load resource: the server responded with a status of 404 (Not Found)`
+    );
+    expect(notFoundMessage.type).toEqual('error');
+    expect(notFoundMessage.timestamp).toBeDefined();
```
Comment (vigneshshanmugam) [inline __tests__/plugins/browser-console.test.ts:90]: nit: we can remove these timestamp checks and keep only required ones.

**Review by vigneshshanmugam — COMMENTED** [review]
Looking good, can you add a test for promise rejection as well.

File: src/reporters/json.ts (hunk 4 above)
Comment (shahzad31) [reply src/reporters/json.ts:492]: write handler will get only text from the error, no? if we want the stack trace implicitely we will have to get it like it's done here...

**Comment by shahzad31** [comment]
> Looking good, can you add a test for promise rejection as well.

@vigneshshanmugam promoise rejections doesnt let the test run. I think current test cover enough scenerios.

**Comment by vigneshshanmugam** [comment]
> @vigneshshanmugam promoise rejections doesnt let the test run. I think current test cover enough scenerios.

Probably its a problem then, The intention of the original issue was to capture unhandled exceptions. Will merge once i test it out.

File: src/reporters/json.ts (hunk 4 above)
Comment (vigneshshanmugam) [reply src/reporters/json.ts:492]: The payload would contain `type and text` as it was previously and you would write `errror` in top level which would be picked up here - https://github.com/elastic/synthetics/blob/8c502cfa027781f4031829d97763ac3ad3496cf4/src/reporters/json.ts#L589

File: __tests__/plugins/browser-console.test.ts (hunk 5 above)
Comment (shahzad31) [reply __tests__/plugins/browser-console.test.ts]: done

File: src/reporters/json.ts (hunk 4 above)
Comment (shahzad31) [reply src/reporters/json.ts:493]: got it, i have pushed the update

**Review by vigneshshanmugam — APPROVED** [review]

**Pull request merged**